CRM/Bsd/Page/BSD.php

```diff
@@ -168,69 +183,109 @@ public function share($param) {
-
     $result = civicrm_api3('Contact', 'get', $contact);
 
-    unset($contact[$apiAddressGet]);
-    unset($contact[$apiGroupContactGet]);
+    if ($result['count'] == 1) {
+      $contact = $this->prepareParamsContact($param, $contact, $result, $result['values'][0]['id']);
+    } elseif ($result['count'] > 1) {
+      $new_contact = $contact;
+      $new_contact['first_name'] = $h->firstname;
+      $new_contact['last_name'] = $h->lastname;
+      $similarity = $this->glueSimilarity($new_contact, $result['values']);
+      unset($new_contact);
+      $contact_id_best = $this->chooseBestContact($similarity);
+      $contact = $this->prepareParamsContact($param, $contact, $result, $contact_id_best);
+    } else {
+      $this->new_contact = true;
+      $contact = $this->prepareParamsContact($param, $contact, $result);
+    }
+
+    CRM_Core_Error::debug_var('$createContact_PARAMS', $contact, false, true);
+    return civicrm_api3('Contact', 'create', $contact);
+
+  }
+
+
+  /**
+   * Preparing params for API Contact.create based on retrieved result.
+   * @param array $param
+   * @param array $contact
+   * @param array $result
+   * @param int $based_on_contact_id
+   *
+   * @return mixed
+   */
+  function prepareParamsContact($param, $contact, $result, $based_on_contact_id = 0) {
+    $h = $param->cons_hash;
+
+    $opt_in_map_group_status = array(
+      0 => 'Added',
+      1 => 'Pending', //default
+    );
+
+    unset($contact[$this->apiAddressGet]);
+    unset($contact[$this->apiGroupContactGet]);
     if ($h->addresses[0]->zip[0] =="[" and $h->addresses[0]->zip[3] == "]") {
-      $contact[$apiAddressCreate] = array(
-       'postal_code' => substr($h->addresses[0]->zip,4),
-       'country' => substr($h->addresses[0]->zip,1,2),
-       'is_primary' => 1,
+      $contact[$this->apiAddressCreate] = array(
+        'postal_code' => substr($h->addresses[0]->zip, 4),
+        'country' => substr($h->addresses[0]->zip, 1, 2),
+        'is_primary' => 1,
       );
     } else {
-      $contact[$apiAddressCreate] = array(
+      $contact[$this->apiAddressCreate] = array(
         'postal_code' => $h->addresses[0]->zip,
         'is_primary' => 1,
       );
     }
-       
-    $opt_in_map_group_status = array(
-      0 => 'Added',
-      1 => 'Pending', //default
-    );
-    CRM_Core_Error::debug_var('$opt_in_map_group_status[$this->opt_in]', $opt_in_map_group_status[$this->opt_in], false, true);
-    if ($result['count'] == 1) {
-      $contact['id'] = $result['values'][0]['id'];
-      if ($result['values'][0][$apiAddressGet]['count'] == 1) {
-        $contact[$apiAddressCreate]['id'] = $result['values'][0]['address_id'];
+
+    $existing_contact = array();
+    if ($based_on_contact_id > 0) {
+      foreach ($result['values'] as $id => $res) {
+        if ($res['id'] == $based_on_contact_id) {
+          $existing_contact = $res;
+        }
+      }
+    }
+
+    if (is_array($existing_contact) && count($existing_contact) > 0) {
```

Comment:
What happen if the contact has more than one address? are you sure you update one instead of adding more?

ideally: check existing addresses: if one has only a zip OR country and it's the same as the one if the form->update, if not, create